Plugins/Public/mobiledocking_plugin/Main.cpp

```diff
@@ -201,252 +82,618 @@ void LoadSettings()
+		dockShipOnCarrier(dockingID, carrierID);
+	}
+}
 
-			// Clear the client out of the mobiledockClients now that it's no longer docked
-			mobiledockClients.erase(client);
+void AddClientToDockQueue(uint dockingID, uint carrierID)
+{
+	const auto& dockMode = mobiledockClients[carrierID].dockMode;
+	if (dockMode == ALLOW_ALL)
+	{
+		StartDockingProcedure(dockingID, carrierID);
+	}
+	else if (dockMode == ALLOW_NONE)
+	{
+		mapPendingDockingRequests[dockingID] = carrierID;
+		PrintUserCmdText(carrierID, L"%s is requesting to dock, authorise with /allowdock", Players.GetActiveCharacterName(dockingID));
+		PrintUserCmdText(dockingID, L"Docking request sent to %s", Players.GetActiveCharacterName(carrierID));
+	}
+	else if (dockMode == ALLOW_GROUP)
+	{
+		list<GROUP_MEMBER> lstGrpMembers;
+		HkGetGroupMembers((const wchar_t*)Players.GetActiveCharacterName(dockingID), lstGrpMembers);
 
-			// Remove charname from carrier's mapDockedShips.
-			mobiledockClients[carrier_client].mapDockedShips.erase(clientName);
-			mobiledockClients[carrier_client].iDockingModulesAvailable++;
+		bool isGroupMember = false;
+		for (auto& member : lstGrpMembers) {
+			if (member.iClientID == carrierID)
+			{
+				isGroupMember = true;
+				break;
+			}
+		}
+		if (isGroupMember)
+		{
+			StartDockingProcedure(dockingID, carrierID);
 		}
-		// If not - redirect to proxy base in carrier's system.
 		else
 		{
-			mobiledockClients[client].carrierSystem = -1;
+			mapPendingDockingRequests[dockingID] = carrierID;
+			PrintUserCmdText(carrierID, L"%s is requesting to dock, authorise with /allowdock", Players.GetActiveCharacterName(dockingID));
+			PrintUserCmdText(dockingID, L"Docking request sent to %s", Players.GetActiveCharacterName(carrierID));
 		}
 	}
 }
 
-bool __stdcall LaunchPosHook(uint space_obj, struct CEqObj &p1, Vector &pos, Matrix &rot, int dock_mode)
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void __stdcall BaseExit(uint iBaseID, uint iClientID)
 {
 	returncode = DEFAULT_RETURNCODE;
 
-	// Redirect the ship to carrier's position. Can bug with POB plugin, changes may be required.
-	if (undockingShip.proxyBaseID == space_obj)
-	{
-		returncode = SKIPPLUGINS_NOFUNCTIONCALL;
-
-		rot = undockingShip.carrierRot;
-		pos = undockingShip.carrierPos;
-		undockingShip.proxyBaseID = 0;
+	mobiledockClients[iClientID].iDockingModulesInstalled = GetInstalledModules(iClientID);
+	
+	if (idToCarrierInfoMap.count(iClientID)) {
+		// Normalize the docking modules available, with the number of people currently docked
+		mobiledockClients[iClientID].iDockingModulesAvailable = (mobiledockClients[iClientID].iDockingModulesInstalled - idToCarrierInfoMap[iClientID]->dockedShipList.size());
 	}
-	return true;
+	else {
+		mobiledockClients[iClientID].iDockingModulesAvailable = mobiledockClients[iClientID].iDockingModulesInstalled;
+	}
+
 }
 
 void __stdcall PlayerLaunch_AFTER(unsigned int ship, unsigned int client)

```

Comment:
Are all Disco hooks required to be __stdcall or is that a hold over from FLHook 1.4 you guys are still doing?

Reply:
I have once forgot to add those descriptors and WEIRD SHIT started happening until I spotted it and added them back. Therefore it stays.